src/state_tracker/state.rs

```diff
@@ -66,77 +66,66 @@ where
                     "Map keys must be strings",
                 ))));
             },
-            (Some(MapValue(label)), List) => {
-                self.state.push(MapKey(Some(label)));
-                if self.state.len() >= self.max_depth {
-                    return self.latch_err(Err(E::from(StructureError::NestingTooDeep)));
-                }
-                self.state.push(Seq);
-            },
-            (Some(MapValue(label)), Dict) => {
-                self.state.push(MapKey(Some(label)));
+            (Some(MapValue(label)), List) | (Some(MapValue(label)), Dict) => {
+                let dummy: &[u8] = &[1];
+                self.state[last_index] = MapKey(Some(std::mem::replace(label, dummy.into())));
                 if self.state.len() >= self.max_depth {
                     return self.latch_err(Err(E::from(StructureError::NestingTooDeep)));
                 }
-                self.state.push(MapKey(None));
+                self.state
+                    .push(if token == &List { Seq } else { MapKey(None) });
             },
-            (Some(oldstate @ MapValue(_)), End) => {
-                self.state.push(oldstate);
+            (Some(_oldstate @ MapValue(_)), End) => {
+                self.state.pop();
                 return self.latch_err(Err(E::from(StructureError::invalid_state(
                     "Missing map value",
                 ))));
             },
             (Some(MapValue(label)), _) => {
-                self.state.push(MapKey(Some(label)));
+                let dummy: &[u8] = &[1];
+                self.state[last_index] = MapKey(Some(std::mem::replace(label, dummy.into())));
             },
-            (oldstate, List) => {
-                if let Some(oldstate) = oldstate {
-                    self.state.push(oldstate);
+            (oldstate, List) | (oldstate, Dict) => {
+                if oldstate.is_none() && 0 < self.state.len() {
+                    self.state.pop();
                 }
                 if self.state.len() >= self.max_depth {
                     return self.latch_err(Err(E::from(StructureError::NestingTooDeep)));
                 }
-                self.state.push(Seq);
+                self.state
+                    .push(if token == &List { Seq } else { MapKey(None) });
             },
-            (oldstate, Dict) => {
-                if let Some(oldstate) = oldstate {
-                    self.state.push(oldstate);
-                }
-
-                if self.state.len() >= self.max_depth {
-                    return self.latch_err(Err(E::from(StructureError::NestingTooDeep)));
-                }
-                self.state.push(MapKey(None));
-            },
-            (oldstate, _) => {
-                if let Some(oldstate) = oldstate {
-                    self.state.push(oldstate);
-                }
+            (oldstate, _) if oldstate.is_none() && 0 < actual_len => {
```

Comment:
again, if oldstate is None, the state stack is empty, so this branch will never be taken.

Reply:
Not necessarily, if oldstate is None, and token is not in the match arms elsewhere, then this branch could execute, but the match arms seems exhaustive so this shouldn't happen..